packages/react-dom/src/client/ReactDOMRoot.js

```diff
@@ -107,17 +107,18 @@ ReactDOMHydrationRoot.prototype.render = ReactDOMRoot.prototype.render =
     }
 
     if (__DEV__) {
-      if (typeof arguments[1] === 'function') {
+      // Avoid GCC optimizations affecting function arity
+      if (arguments.length >= 2 && typeof arguments[1] === 'function') {
         console.error(
           'does not support the second callback argument. ' +
             'To execute a side effect after rendering, declare it in a component body with useEffect().',
         );
-      } else if (isValidContainer(arguments[1])) {
+      } else if (arguments.length >= 2 && isValidContainer(arguments[1])) {
```

Comment:
Instead of checking for `arguments.length >=2` in every if block, you can wrap it to the parent if block. This way the code will be faster by avoiding checking the arguments. length everytime